qutebrowser/browser/webengine/notification.py

```diff
@@ -0,0 +1,218 @@
+        bus = QDBusConnection.sessionBus()
+        if not bus.isConnected():
+            raise DBusException("Failed to connect to DBus session bus")
+
+        service = self.TEST_SERVICE if test_service else self.SERVICE
+
+        self.interface = QDBusInterface(
+            service,
+            self.PATH,
+            self.INTERFACE,
+            bus,
+        )
+
+        bus.connect(
+            service,
+            self.PATH,
+            self.INTERFACE,
+            "NotificationClosed",
+            self._handle_close
+        )
+
+        if not self.interface:
+            raise DBusException("Could not construct a DBus interface")
+
+    def install(self, profile: "QWebEngineProfile") -> None:
+        """Sets the profile to use the manager as the presenter."""
+        # WORKAROUND for
+        # https://www.riverbankcomputing.com/pipermail/pyqt/2020-May/042916.html
+        if PYQT_VERSION < 0x050F00:  # PyQt 5.15
+            # PyQtWebEngine unrefs the callback after it's called, for some
+            # reason.  So we call setNotificationPresenter again to *increase*
+            # its refcount to prevent it from getting GC'd. Otherwise, random
+            # methods start getting called with the notification as `self`, or
+            # segfaults happen, or other badness.
+            def _present_and_reset(qt_notification: "QWebEngineNotification") \
+                    -> None:
+                profile.setNotificationPresenter(_present_and_reset)
+                self._present(qt_notification)
+            profile.setNotificationPresenter(_present_and_reset)
+        else:
+            profile.setNotificationPresenter(self._present)
+
+    def _present(self, qt_notification: "QWebEngineNotification") -> None:
+        """Shows a notification over DBus.
+
+        This should *not* be directly passed to setNotificationPresenter on
+        PyQtWebEngine < 5.15 because of a bug in the PyQtWebEngine bindings.
+        """
+        # notification id 0 means 'assign us the ID'. We can't just pass 0
+        # because it won't get sent as the right type.
+        zero = QVariant(0)
+        zero.convert(QVariant.UInt)
+
+        actions_list = QDBusArgument([], QMetaType.QStringList)
+
+        qt_notification.show()
+        hints = {
+            # Include the origin in case the user wants to do different things
+            # with different origin's notifications.
+            "x-qutebrowser-origin": qt_notification.origin().toDisplayString()
+        }  # type: typing.Dict[str, typing.Any]
+        if not qt_notification.icon().isNull():
+            hints["image-data"] = self._convert_image(qt_notification.icon())
+
+        reply = self.interface.call(
+            QDBus.BlockWithGui,
+            "Notify",
+            "qutebrowser",  # application name
+            zero,  # notification id
+            "qutebrowser",  # icon
+            qt_notification.title(),
+            qt_notification.message(),
```

Comment:
The Desktop Notifications Specification allows [simple Markup](https://developer.gnome.org/notification-spec/#markup) as part of messages. What should happen if the website sends such markup as part of a notification? What are the security implications of letting websites show messages with markup? What do other browsers using XDG notifications (if any?) do about that?

Reply:
Right, I forgot about this. We want to never send markup, since it's not going to work cross-platform and the HTML spec doesn't say anything about any kind of markup. So we need to entity-escape it it if the server says it supports markup (there's a capabilities API), but not if the server says it doesn't. We can do that check as part of the Presenter's initialization and then cache the result forever.